src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +128,318 @@ object RuleUtils {
       None // logicalPlan is non-linear or it has no LogicalRelation.
     }
   }
+
+  /**
+   * Transform the current plan to utilize the given index.
+   *
+   * The transformed plan reads the given index data rather than original source files.
+   * If HybridScan is enabled, additional logical plans for the appended source files would be
+   * constructed and merged with the transformed plan with the index.
+   *
+   * Pre-requisites
+   * - We know for sure the index which can be used to replace the plan.
+   * - The plan should be linear and include 1 LogicalRelation.
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Replacement plan.
+   */
+  def transformPlanToUseIndex(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    val transformed = if (HyperspaceConf.hybridScanEnabled(spark)) {
+      transformPlanToUseHybridIndexDataScan(spark, index, plan, useBucketSpec)
+    } else {
+      transformPlanToUsePureIndexScan(spark, index, plan, useBucketSpec)
+    }
+    assert(!transformed.equals(plan))
+    transformed
+  }
+
+  /**
+   * Transform the current plan to utilize index
+   * The transformed plan reads data from indexes instead of the source relations.
+   * Bucketing information of the index is retained if useBucketSpec is true.
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Transformed logical plan that leverages an index.
+   */
+  private def transformPlanToUsePureIndexScan(
```

Comment:
(just a naming suggestion, feel free to ignore it if you disagree)
rename `transformPlanToUsePureIndexScan` -> `transformPlanToUseIndexOnlyScan`

Reply:
The name was also suggested.  I think `PureIndexScan` is also good naming :)